Hostname related changes

sandeen opened the pull request:
I have one openevse in "production" and one on the desk; having the OTA target push to "openevse" is ... dangerous when there are 2 hosts running with the same name.

The first commit allows the hostname to be set during build time via `-DHOSTNAME`, and changes it name to "openevse-ota" (we could call it -devel, or whatever you like?)

The second commit changes the authentication realm from "asyncesp" to whatever the wifi hostname has been set to (openevse, openevse-ota, etc), for consistency.

(Eventually an option on a configuration tab to choose the hostname might be nice for anyone with more than one openevse in production)

jeremypoulter commented on platformio.ini:
```diff
@@ -37,5 +37,5 @@ platform = espressif8266
 lib_deps = ${common.lib_deps}
-src_build_flags = ${common.version} -DENABLE_OTA -DWIFI_LED=0 -DENABLE_DEBUG
-upload_port = openevse.local
+src_build_flags = ${common.version} -DENABLE_OTA -DWIFI_LED=0 -DENABLE_DEBUG -DHOSTNAME=openevse-ota
```
Not totally sure I like this, really what we need is some way to set this locally outside of what is checked in, or ideally, as you say save it in the settings.

You can use ${env.XXXX} to pull in system [environment vars](http://docs.platformio.org/en/latest/projectconf.html#dynamic-variables) but I don't see a way of setting a default when not set. Another option may be to use a [script ](http://docs.platformio.org/en/latest/projectconf.html#dynamic-variables) to read in the host name from a file, giving a default if not there?

sandeen replied:
I looked at the script option, but that depends heavily on the OS (at least, Windows can't run a bash script) ...

jeremypoulter commented on src/wifi.cpp:
```diff
@@ -16,7 +16,11 @@ IPAddress apIP(192, 168, 4, 1);
 IPAddress netMsk(255, 255, 255, 0);
 
 // hostname for mDNS. Should work at least on windows. Try http://openevse or http://openevse.local
+#ifdef HOSTNAME
```
Personally I prefer to always have 
HOSTNAME set using;
```
#ifndef HOSTNAME
#define HOSTNAME openevse
#endif
```
somewhere near the top of the file, see [wifi.cpp](https://github.com/OpenEVSE/ESP8266_WiFi_v2.x/blob/master/src/wifi.cpp#L30)

sandeen replied:
Good point, thanks.

sandeen commented:
Ok, I rebased this branch, and now the `openevse_ota` pio environment will set the hostname to, and upload to, whatever is set in the `ESP_WIFI_HOSTNAME ` environment variable.  I added the hostname to the startup debug text as well.

Couldn't find a clean cross-platform way to make the pio environment pick a default, but setting the env var in my .bashrc worksforme.  (If `ESP_WIFI_HOSTNAME` is unset, the code will default to "openevse" as a hostname, but there won't be any default upload_port, I'm afraid)

Fixed up the `#ifdef` stuff too.

glynhudson commented:
Would it be possible to pull the latest changes into this PR? Thanks, I'll get it merged asap. Cheers,

sandeen commented:
@jeremypoulter how do you feel about the changes at this point, using env. vars in platformio.ini?

jeremypoulter commented:
I think the env will work. Does it default to openevse if no env is set?

sandeen commented:
The binary should default to "openevse" but I could not find a way to create a default OTA upload target.

sandeen commented:
I'll just close this one for now, it's getting stale and I guess there are still oddities around how the default name gets set.  I'll just name my production unit differently so there's no clash.
